segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableEntryDeltaIterator.java

```diff
@@ -0,0 +1,193 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.tables;
+
+import io.pravega.common.TimeoutTimer;
+import io.pravega.common.util.AsyncIterator;
+import io.pravega.common.util.BufferView;
+import io.pravega.segmentstore.contracts.ReadResult;
+import io.pravega.segmentstore.contracts.tables.TableEntry;
+import io.pravega.segmentstore.server.DirectSegmentAccess;
+
+import java.io.IOException;
+import java.time.Duration;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import io.pravega.segmentstore.server.reading.AsyncReadResultProcessor;
+import lombok.Builder;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+/**
+ * Iterates through a {@link DirectSegmentAccess}, deserializing {@link TableEntry} from a {@link BufferView} in linear fashion.
+ *
+ * @param <T> Type of the final, converted result.
+ */
+@Slf4j
+@ThreadSafe
+@Builder
+class TableEntryDeltaIterator<T> implements AsyncIterator<T> {
+    //region Members
+
+    // The maximum size (in bytes) of each read to perform on the segment.
+    private static final int MAX_READ_SIZE = 2 * 1024 * 1024;
+
+    private final DirectSegmentAccess segment;
+    // The offset to being iteration at.
+    private final long startOffset;
+    // Maximum length of the TableSegment we want to read until.
+    private final int maxLength;
+    private final boolean shouldClear;
+    private final Duration fetchTimeout;
+    private final EntrySerializer entrySerializer;
+    private final ConvertResult<T> resultConverter;
+    private final Executor executor;
+
+    @GuardedBy("this")
+    private Iterator<Map.Entry<DeltaIteratorState, TableEntry>> currentEntry;
+    @GuardedBy("this")
+    private long currentBatchOffset;
+
+    //endregion
+
+    //region AsyncIterator Implementation
+
+    @Override
+    public CompletableFuture<T> getNext() {
```

Comment:
It says this is thread safe, but I am not clear on the semantics of this method being called in parallel. 
In particular `getNextEntry` could have two calls to `getNextEntryFromBatch` in parallel both get null and hence end up calling `fetchNextTableEntriesBatch` twice. When if those calls had been ordered it would have only resulted to one call to `fetchNextTableEntriesBatch`

Reply:
I see, good catch. It would seem the only conflict that can occur is that multiple calls to `getNextEntry` can return `null`, but otherwise concurrent calls won't point to the same non-null `currentEntry`. So I see two solutions: call `if (this.currentEntry != null) return CompletableFuture.completedFuture(null);` from within `fetchNextTableEntriesBatch` or make `getNextEntry` synchronized. I'm not sure the former provides any substantial improvement.